homeassistant/components/systemmonitor/coordinator.py

```diff
@@ -1,131 +1,129 @@
 
-from datetime import timedelta
-from functools import cache
 import logging
 import os
-from typing import Literal
 
 import psutil
 from psutil._common import sdiskusage, shwtemp, snetio, snicaddr, sswap
 from psutil._pslinux import svmem
 
 from homeassistant.core import HomeAssistant
 from homeassistant.helpers.entity_component import DEFAULT_SCAN_INTERVAL
-from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
 
 from .const import DOMAIN
 from .sensor import CPU_SENSOR_PREFIXES
 
 _LOGGER = logging.getLogger(__name__)
 
 
-class SystemMonitorDiskCoordinator(
-    DataUpdateCoordinator[
-        (
-            sdiskusage
-            | sswap
-            | svmem
-            | dict[str, snetio]
-            | dict[str, list[snicaddr]]
-            | tuple[float, float, float]
-            | float
-            | None
-        )
-    ]
-):
-    """A System monitor Data Update Coordinator."""
-
-    def __init__(
-        self,
-        hass: HomeAssistant,
-        type_: Literal["disk", "swap", "memory", "netio", "netaddr", "load", "cputemp"],
-        argument: str,
-    ) -> None:
+class SystemMonitorDiskCoordinator(DataUpdateCoordinator[sdiskusage]):
+    """A System monitor Disk Data Update Coordinator."""
+
+    def __init__(self, hass: HomeAssistant, argument: str) -> None:
         """Initialize the coordinator."""
         super().__init__(
-            hass,
-            _LOGGER,
-            name=DOMAIN,
-            update_interval=timedelta(seconds=DEFAULT_SCAN_INTERVAL.seconds),
+            hass, _LOGGER, name=DOMAIN, update_interval=DEFAULT_SCAN_INTERVAL
         )
-        self._type = type_
         self._argument = argument
 
-    async def _async_update_data(
-        self,
-    ) -> (
-        sdiskusage
-        | sswap
-        | svmem
-        | dict[str, snetio]
-        | dict[str, list[snicaddr]]
-        | tuple[float, float, float]
-        | float
-        | None
-    ):
+    async def _async_update_data(self) -> sdiskusage:
         """Fetch data."""
+        return psutil.disk_usage(self._argument)
 
-        if self._type == "disk":
-            return _disk_usage(self._argument)
-        if self._type == "swap":
-            return _swap_memory()
-        if self._type == "memory":
-            return _virtual_memory()
-        if self._type == "netio":
-            return _net_io_counters()
-        if self._type == "netaddr":
-            return _net_if_addrs()
-        if self._type == "load":
-            return _getloadavg()
-        if self._type == "cputemp":
-            return _read_cpu_temperature()
 
+class SystemMonitorSwapCoordinator(DataUpdateCoordinator[sswap]):
+    """A System monitor Swap Data Update Coordinator."""
 
-@cache
-def _disk_usage(path: str) -> sdiskusage:
-    return psutil.disk_usage(path)
+    def __init__(self, hass: HomeAssistant) -> None:
```

Comment:
I think we can make a base class with the default init method and keep it generic via a type var for the coordinator data value. Then we don't need to repeat the init method in every coordinator, unless we need to customize the init method.

```py
_dataT = TypeVar("dataT")


class MonitorCoordinator(DataUpdateCoordinator[_dataT]):

    def __init__(self, hass: HomeAssistant) -> None:
        super().__init__(
            hass, _LOGGER, name=DOMAIN, update_interval=DEFAULT_SCAN_INTERVAL
        )

    async def _async_update_data(self) -> _dataT:
        return await self.update_data()

    @abstractmethod
    async def update_data(self) -> _dataT:

Reply:
Need to split it up even further than current so makes sense.
We can add the argument always to the base class so then we never need to repeat `__init__`